### PR TITLE
Make vCOW swappable

### PR DESCRIPTION
**Pull request by fedgiac**
Creates a script that generates the DAO proposal for making vCOW swappable.

For now, it only handles mainnet, but it will be extended to handle Gnosis Chain.

### Test Plan

New unit tests. Also, try out the script as in the readme. I simulated the resulting transaction on Tenderly and it does execute the transfer, but I'm not able to get a link I can use to share, so you'd have to try yourself.

### CODE REVIEW

File: test/make-swappable-proposal.test.ts
Comment (josojo) [inline test/make-swappable-proposal.test.ts:32]: Unwanted?

**Review by josojo — APPROVED** [review]
Nice!

I did the successful tenderly simulation:
Can you access this link?
https://dashboard.tenderly.co/gp-v2/staging/simulator/a5efffc7-fc0b-4c7e-847b-9a99360d3462

File: example/test-make-vcow-swappable.json
Comment (fedgiac) [reply example/test-make-vcow-swappable.json:2]: To me the grouping is handy as mainnet contracts will be written as `mainnet.vCow`, compared to `mainnetVcow`. So the choice is between:

```
{
"mainnet": {
    "cowToken": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
    "virtualCowToken": "0xd057b63f5e69cf1b929b356b579cba08d7688048",
    "atoms": "123456000000000000000000",
    "multisend": "0x8d29be29923b68abfdd21e541b9374737b49cdad"
  },
"gnosisChain": {
    "cowToken": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
    "virtualCowToken": "0xd057b63f5e69cf1b929b356b579cba08d7688048",
    "atoms": "123456000000000000000000",
    "multisend": "0x8d29be29923b68abfdd21e541b9374737b49cdad"
  }
}
```

and:

```
{
  "mainnetCowToken": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
  "mainnetVirtualCowToken": "0xd057b63f5e69cf1b929b356b579cba08d7688048",
  "mainnetAtoms": "123456000000000000000000",
  "mainnetMultisend": "0x8d29be29923b68abfdd21e541b9374737b49cdad"
  "gcCowToken": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
  "gcVirtualCowToken": "0xd057b63f5e69cf1b929b356b579cba08d7688048",
  "gcAtoms": "123456000000000000000000",
}
```

I personally prefer the first, and I think it might be better to clearly distinguish mainnet and GC, but I'm also ok with the second.

**Comment by fedgiac** [comment]
Thanks for the link, it works for me!

File: example/test-make-vcow-swappable.json
Comment (josojo) [reply example/test-make-vcow-swappable.json:2]: yes, the grouping is valuable. I was thinking about such a config: 
```
{
  "VcowContractOnMainnet": {
      "virtualCowToken": "0xd057b63f5e69cf1b929b356b579cba08d7688048",
      "atomsToTransfer": "123456000000000000000000",
    },
  "VcowContractOnGC": {
      "virtualCowToken": "different address",
      "atomsToTransfer": "123456000000000000000000",
    }
    "cowToken": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
    "multisend": "0x8d29be29923b68abfdd21e541b9374737b49cdad"
}
```
Sorry for not making it more clear before.

But I am also totally with your current version. Feel free to ignore all my suggestions

Comment (fedgiac) [reply example/test-make-vcow-swappable.json]: I see, at first I did the splitting in two because I want to clearly point out that also the multisig and the cow token belong to mainnet, but maybe that's not really helpful as you mention.

For this PR, in hindsight I shouldn't have included "mainnet" as there is only mainnet available so it should be implied. So for now I just removed the `mainnet` wrapper (and renamed `atoms` to `atomsToTransfer`). In this way this discussion should not be blocking this PR. Anyway I'm soon publishing the PR for the bridging on GC, and I think this discussion would be more fitting there.

On the topic, I'm wondering if actually using `mainnet` and `GC` is actually a good thing, since in a way it hardcodes the networks we want to use. I'm now thinking about something like this now:

```
{
  "bridged": {
    "virtualCowToken": "0xd057b63f5e69cf1b929b356b579cba08d7688048",
    "atomsToTransfer": "123456000000000000000000",
  },
  "virtualCowToken": "different address",
  "atomsToTransfer": "123456000000000000000000",
  "cowToken": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
  "multisend": "0x8d29be29923b68abfdd21e541b9374737b49cdad"
  "multiTokenMediator": "0x88ad09518695c6c3712AC10a214bE5109a655671",
}
```
The implication being that the lower level is the network where the script is run, while the `bridged` level is the network we bridge to.

Comment (josojo) [reply example/test-make-vcow-swappable.json]: You are right. This way it makes the most sense!

**Review by josojo — APPROVED** [review]

**Review by fleupold — COMMENTED** [review]
This is a nice script! I'm looking into creating the transactions for the solver reward proposal and am wondering if we should have a separate repository for CiP related code so that for each CiP we can generate such a script. This would allow easy review + sharing with the community.

**Comment by fedgiac** [comment]
I think we should have a separate repository. The token is "static" but we could have new proposals at any time. As you mention, I think it would be good for transparency. I will start looking into it tomorrow and wait before merging.

**Comment by fedgiac** [comment]
Since we already have some proposal-related code in this repo, I'm merging all open PRs and using them as the base to copy the code of the new repo. So that there still is a trail of PRs. I'll delete everything from this repo if we are satisfied with the new one.

**Comment by fedgiac** [comment]
I created a proposal repo including only the code for this proposal. What do you think @fleupold @josojo?
https://github.com/fedgiac/cow-proposals

I plan to remove some boilerplate in the future. Also, I left in some code that was used by the deployment proposal, even if I decided to leave the deployment proposal here. This is because I expect the tooling to be useful in the future.

If you like it, I'll transfer it to Gnosis and make all remaining PRs about this proposal on top of it instead.

**Comment by fleupold** [comment]
Yes, I think this looks great and will provide a proper engineering framework around proposal creation (review, reproducibility, unit tests, simulation, etc).

**Pull request merged**
